stacks-signer/src/chainstate.rs

```diff
@@ -0,0 +1,376 @@
+            if !confirms_latest_in_tenure {
+                return Ok(false);
+            }
+        }
+
+        if let Some(tenure_extend) = block.get_tenure_extend_tx_payload() {
+            // in tenure extends, we need to check:
+            // (1) if this is the most recent sortition, an extend is allowed if it changes the burnchain view
+            // (2) if this is the most recent sortition, an extend is allowed if enough time has passed to refresh the block limit
+            let changed_burn_view =
+                tenure_extend.burn_view_consensus_hash != proposed_by.state().consensus_hash;
+            let enough_time_passed = Self::tenure_time_passed_block_lim()?;
+            if !changed_burn_view && !enough_time_passed {
+                warn!(
+                    "Miner block proposal contains a tenure extend, but the burnchain view has not changed and enough time has not passed to refresh the block limit. Considering proposal invalid.";
+                    "proposed_block_consensus_hash" => %block.header.consensus_hash,
+                    "proposed_block_signer_sighash" => %block.header.signer_signature_hash(),
+                );
+                return Ok(false);
+            }
+        }
+
+        Ok(true)
+    }
+
+    fn check_parent_tenure_choice(
+        sortition_state: &SortitionState,
+        block: &NakamotoBlock,
+        client: &StacksClient,
+    ) -> Result<bool, ClientError> {
+        // if the parent tenure is the last sortition, it is a valid choice.
+        // if the parent tenure is a reorg, then all of the reorged sortitions
+        //  must either have produced zero blocks _or_ produced their first block
+        //  very close to the burn block transition.
+        if sortition_state.prior_sortition == sortition_state.parent_tenure_id {
+            return Ok(true);
+        }
+        info!(
+            "Most recent miner's tenure does not build off the prior sortition, checking if this is valid behavior";
+            "proposed_block_consensus_hash" => %block.header.consensus_hash,
+            "proposed_block_signer_sighash" => %block.header.signer_signature_hash(),
+        );
+
+        let tenures_reorged = client.get_tenure_forking_info(
+            &sortition_state.parent_tenure_id,
+            &sortition_state.prior_sortition,
+        )?;
+        if tenures_reorged.is_empty() {
+            warn!("Miner is not building off of most recent tenure, but stacks node was unable to return information about the relevant sortitions. Marking miner invalid.";
+                    "proposed_block_consensus_hash" => %block.header.consensus_hash,
+                    "proposed_block_signer_sighash" => %block.header.signer_signature_hash(),
+            );
+            return Ok(false);
+        }
+        for tenure in tenures_reorged.iter() {
+            if tenure.first_block_mined.is_some() {
+                // TODO: must check if the first block was poorly timed.
```

Comment:
Is there an issue open for this?

Reply:
Added https://github.com/stacks-network/stacks-core/issues/4844